File: cpp/taskflow/matmul.cpp
// An implementation of recursive matrix multiplication

// Adapted from
// https://github.com/mtmucha/coros/blob/main/benchmarks/coros_mat.h

// Original author: mtmucha
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE or copy at http://www.boost.org/LICENSE_1_0.txt)

#include "matmul.hpp"
#include <taskflow/algorithm/for_each.hpp>
#include <taskflow/taskflow.hpp>

#include <chrono>
#include <cstdio>
#include <exception>
#include <vector>

static int thread_count = std::thread::hardware_concurrency() / 2;
static const size_t iter_count = 1;

void matmul(tf::Runtime& rt, int* a, int* b, int* c, int n, int N) {
  if (n <= 32) {
    // Base case: Use simple triple-loop multiplication for small matrices
    matmul_small(a, b, c, n, N);
  } else {
    // Recursive case: Divide the matrices into 4 submatrices and multiply them
    int k = n / 2;

    // Split the execution into 2 sections to ensure output locations are not
    // written in parallel
    rt.silent_async([=](tf::Runtime& s) { matmul(s, a, b, c, k, N); });
    rt.silent_async([=](tf::Runtime& s) { matmul(s, a, b + k, c + k, k, N); });
    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k * N, b, c + k * N, k, N);
    });
    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k * N, b + k, c + k * N + k, k, N);
    });
    rt.corun_all();


    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k, b + k * N, c, k, N);
    });
    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k, b + k * N + k, c + k, k, N);
    });
    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k * N + k, b + k * N, c + k * N, k, N);
    });
    rt.silent_async([=](tf::Runtime& s) {
      matmul(s, a + k * N + k, b + k * N + k, c + k * N + k, k, N);
    });
    rt.corun_all();
  }
}

std::vector<int> run_matmul(tf::Executor& executor, int N) {
  std::vector<int> A(N * N, 1);
  std::vector<int> B(N * N, 1);
  std::vector<int> C(N * N, 0);

  int* a = A.data();
  int* b = B.data();
  int* c = C.data();

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      a[i * N + j] = 1;
      b[i * N + j] = 1;
      c[i * N + j] = 0;
    }
  }

  tf::Taskflow taskflow;
  taskflow.emplace([=](tf::Runtime& rt) { matmul(rt, a, b, c, N, N); });
  executor.run(taskflow).wait();
  return C;
}

void validate_result(std::vector<int>& C, int N) {
  std::atomic_thread_fence(std::memory_order_seq_cst);
  int* c = C.data();
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      auto res = c[i * N + j];
      if (res != N) {
        std::printf(
          "Wrong result at (%d,%d) : %d. expected %d\n", i, j, res, N
        );
        std::fflush(stdout);
        std::terminate();
      }
    }
  }
}

void run_one(tf::Executor& executor, int N) {
  auto startTime = std::chrono::high_resolution_clock::now();
  std::vector<int> result = run_matmul(executor, N);
  auto endTime = std::chrono::high_resolution_clock::now();
  validate_result(result, N);
  auto totalTimeUs =
    std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
  std::printf("  - matrix_size: %d\n", N);
  std::printf("    duration: %zu us\n", totalTimeUs.count());
}

int main(int argc, char* argv[]) {
  if (argc != 2) {
    printf("Usage: matmul <matrix size (power of 2)>\n");
    exit(0);
  }
  int n = atoi(argv[1]);
  std::printf("threads: %d\n", thread_count);
  tf::Executor executor(thread_count);

  run_matmul(executor, n); // warmup

  std::printf("runs:\n");

  run_one(executor, n);
}

File: rust/smol/src/bin/matmul.rs
// Recursive divide-and-conquer matrix multiplication.
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/matmul.cpp
// base case: ../../../../cpp/2common/matmul.hpp
//
// The matrices are split into quadrants recursively down to a 32x32 base case.
// The 8 sub-multiplications are run as two sequential groups of 4 parallel
// `Executor::spawn`s so that no output tile is written by two tasks at once.
//
// Rust's borrow checker cannot see that the parallel tasks write to disjoint
// regions of C, so - exactly like the C++ version's `int*` arithmetic - we use
// raw pointers. `Mat` bundles them into a `Send` value that can cross the
// spawn boundary. This is sound because: the tasks only read A/B and write
// disjoint sub-tiles of C, and the owning `Vec`s outlive the blocking call.

use smol::{future, Executor, Task};
use smol_bench::{default_threads, init_executor, peak_memory_usage, BoxFut};
use std::time::Instant;

#[derive(Clone, Copy)]
struct Mat {
    a: *const i32,
    b: *const i32,
    c: *mut i32,
    n: usize,
    nn: usize, // stride (the full matrix dimension N)
}
// SAFETY: each spawned task operates on disjoint sub-tiles of C and read-only A/B.
unsafe impl Send for Mat {}

unsafe fn matmul_small(m: Mat) {
    let Mat { a, b, c, n, nn } = m;
    for i in 0..n {
        for k in 0..n {
            for j in 0..n {
                *c.add(i * nn + j) += *a.add(i * nn + k) * *b.add(k * nn + j);
            }
        }
    }
}

// Compute the two groups of four sub-multiplications. Kept as a plain (non-async)
// function so the raw pointers never live across an `.await`; only the `Send`
// `Mat` values cross the spawn/await boundary.
fn split_quadrants(m: Mat) -> ([Mat; 4], [Mat; 4]) {
    let k = m.n / 2;
    let nn = m.nn;
    let (a, b, c) = (m.a, m.b, m.c);
    unsafe {
        // Group 1: the four products that only touch the "upper-left" halves.
        let group1 = [
            Mat { a, b, c, n: k, nn },
            Mat { a, b: b.add(k), c: c.add(k), n: k, nn },
            Mat { a: a.add(k * nn), b, c: c.add(k * nn), n: k, nn },
            Mat { a: a.add(k * nn), b: b.add(k), c: c.add(k * nn + k), n: k, nn },
        ];
        // Group 2: the accumulating products into the same tiles.
        let group2 = [
            Mat { a: a.add(k), b: b.add(k * nn), c, n: k, nn },
            Mat { a: a.add(k), b: b.add(k * nn + k), c: c.add(k), n: k, nn },
            Mat { a: a.add(k * nn + k), b: b.add(k * nn), c: c.add(k * nn), n: k, nn },
            Mat { a: a.add(k * nn + k), b: b.add(k * nn + k), c: c.add(k * nn + k), n: k, nn },
        ];
        (group1, group2)
    }
}

fn matmul(ex: &'static Executor<'static>, m: Mat) -> BoxFut<()> {
    Box::pin(async move {
        if m.n <= 32 {
            unsafe { matmul_small(m) };
            return;
        }
        let (group1, group2) = split_quadrants(m);
        run_group(ex, group1).await;
        run_group(ex, group2).await;
    })
}

async fn run_group(ex: &'static Executor<'static>, group: [Mat; 4]) {
    let children: Vec<Task<()>> = group.into_iter().map(|m| ex.spawn(matmul(ex, m))).collect();
    for c in children {
        c.await;
    }
}

fn run_matmul(ex: &'static Executor<'static>, n: usize) -> Vec<i32> {
    let a = vec![1i32; n * n];
    let b = vec![1i32; n * n];
    let mut c = vec![0i32; n * n];

    let root = Mat {
        a: a.as_ptr(),
        b: b.as_ptr(),
        c: c.as_mut_ptr(),
        n,
        nn: n,
    };
    future::block_on(ex.spawn(matmul(ex, root)));
    c
}

fn validate(c: &[i32], n: usize) {
    for i in 0..n {
        for j in 0..n {
            let res = c[i * n + j];
            if res != n as i32 {
                eprintln!("Wrong result at ({},{}) : {}. expected {}", i, j, res, n);
                std::process::exit(1);
            }
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        println!("Usage: matmul <matrix size (power of 2)>");
        return;
    }
    let n: usize = args[1].parse().expect("matrix size");
    let thread_count = if args.len() > 2 {
        args[2].parse().expect("thread count")
    } else {
        default_threads()
    };

    let ex = init_executor(thread_count);
    println!("threads: {}", thread_count);

    let _ = run_matmul(ex, n); // warmup

    println!("runs:");
    let start = Instant::now();
    let result = run_matmul(ex, n);
    let dur = start.elapsed();
    validate(&result, n);

    println!("  - matrix_size: {}", n);
    println!("    duration: {} us", dur.as_micros());
    println!("    max_rss: {} KiB", peak_memory_usage());
}
